### PR TITLE
Various improvements to the FileFormats.MPS subpackage

### LINKED ISSUE #1106: MPS code is not deterministic
I haven't tried to create examples, but the use of `Dict` at

https://github.com/jump-dev/MathOptInterface.jl/blob/535e7f0e07e5a56213f2a6b5272972e3503b91d0/src/FileFormats/MPS/MPS.jl#L465-L473

likely makes `copy_to` not deterministic, e.g.,

https://github.com/jump-dev/MathOptInterface.jl/blob/535e7f0e07e5a56213f2a6b5272972e3503b91d0/src/FileFormats/MPS/MPS.jl#L538

https://github.com/jump-dev/MathOptInterface.jl/blob/535e7f0e07e5a56213f2a6b5272972e3503b91d0/src/FileFormats/MPS/MPS.jl#L558

Same issue in the ordering of variables when the problem is written out:

https://github.com/jump-dev/MathOptInterface.jl/blob/535e7f0e07e5a56213f2a6b5272972e3503b91d0/src/FileFormats/MPS/MPS.jl#L169

https://github.com/jump-dev/MathOptInterface.jl/blob/535e7f0e07e5a56213f2a6b5272972e3503b91d0/src/FileFormats/MPS/MPS.jl#L195

https://github.com/jump-dev/MathOptInterface.jl/blob/535e7f0e07e5a56213f2a6b5272972e3503b91d0/src/FileFormats/MPS/MPS.jl#L331

https://github.com/jump-dev/MathOptInterface.jl/blob/535e7f0e07e5a56213f2a6b5272972e3503b91d0/src/FileFormats/MPS/MPS.jl#L344

Unless we have a good reason not to, variables should keep their ordering when you round-trip read and write to an MPS file. The ordering of variables makes a huge difference for MIP solvers. Ideally the same would hold for constraints, although that's a bit more complex because of how we represent them in MOI.

### PR DESCRIPTION
**Pull request by odow**
- Function barriers for type-unstable hot loops
- Deterministic ordering of rows/columns (Closes #1106)
- Remove TempRow and TempColumn in favor of vectors
- Use Int instead of String in coef matrix for lower memory usage

<s>Will post before/after benchmarks reading MIPLIB soon.</s>

Still todo:
- [x] Some instances of non-determinisim

### CODE REVIEW

**Comment by odow** [comment]
tl;dr: 5-10x reduction in allocations. 50% faster.

I picked 10 files from MIPLIB. Then 
```Julia
using MathOptInterface
const MOI = MathOptInterface

function benchmark(filename)
    println("Benchmarking $(filename)")
    src = MOI.FileFormats.Model(format = MOI.FileFormats.FORMAT_MPS)
    dest = MOI.FileFormats.Model(format = MOI.FileFormats.FORMAT_MPS)
    @time MOI.read_from_file(src, filename)
    @time MOI.copy_to(dest, src)
    @time MOI.write_to_file(dest, "out_" * filename)
    return
end

for filename in readdir(".")
    if !endswith(filename, ".mps.gz") || startswith(filename, "out_")
        continue
    end
    benchmark(filename)
end
```

MOI v0.9.14:
```
Benchmarking bab2.mps.gz
 24.295869 seconds (471.76 M allocations: 15.754 GiB, 34.36% gc time)
  3.650245 seconds (12.43 M allocations: 716.467 MiB, 7.72% gc time)
  7.294376 seconds (22.48 M allocations: 890.650 MiB, 7.88% gc time)
Benchmarking irish-electricity.mps.gz
  3.451068 seconds (39.29 M allocations: 1.682 GiB, 30.16% gc time)
  1.904308 seconds (6.90 M allocations: 366.925 MiB, 13.70% gc time)
  2.446018 seconds (6.44 M allocations: 275.112 MiB, 12.11% gc time)
Benchmarking n2seq36q.mps.gz
  2.918412 seconds (89.87 M allocations: 2.881 GiB, 23.37% gc time)
  0.026641 seconds (101.24 k allocations: 17.046 MiB)
  0.463076 seconds (1.26 M allocations: 58.136 MiB, 2.32% gc time)
Benchmarking rocI-4-11.mps.gz
  0.490525 seconds (9.32 M allocations: 325.857 MiB, 36.59% gc time)
  0.972306 seconds (3.82 M allocations: 192.362 MiB, 3.76% gc time)
  0.182439 seconds (616.84 k allocations: 26.949 MiB, 6.83% gc time)
Benchmarking rocII-5-11.mps.gz
  3.469787 seconds (75.90 M allocations: 2.557 GiB, 28.02% gc time)
  0.064427 seconds (391.00 k allocations: 30.660 MiB)
  0.718641 seconds (1.67 M allocations: 72.313 MiB, 7.65% gc time)
Benchmarking roi5alpha10n8.mps.gz
 25.704327 seconds (451.60 M allocations: 15.394 GiB, 38.55% gc time)
  0.240933 seconds (463.35 k allocations: 119.877 MiB)
  6.465842 seconds (10.38 M allocations: 479.186 MiB, 16.39% gc time)
Benchmarking supportcase42.mps.gz
  3.317878 seconds (73.48 M allocations: 2.573 GiB, 26.60% gc time)
  0.060980 seconds (306.58 k allocations: 33.658 MiB)
  0.978155 seconds (2.75 M allocations: 110.861 MiB, 4.04% gc time)
Benchmarking timtab1.mps.gz
  0.025012 seconds (244.89 k allocations: 8.387 MiB, 67.91% gc time)
  0.000744 seconds (4.72 k allocations: 462.641 KiB)
  0.005391 seconds (28.49 k allocations: 1.395 MiB)
Benchmarking triptim1.mps.gz
  4.670464 seconds (116.63 M allocations: 3.906 GiB, 26.67% gc time)
  0.127001 seconds (365.36 k allocations: 46.391 MiB, 13.89% gc time)
  1.217245 seconds (2.53 M allocations: 111.423 MiB, 5.14% gc time)
Benchmarking uccase12.mps.gz
  6.530908 seconds (152.37 M allocations: 5.062 GiB, 31.87% gc time)
  0.489473 seconds (1.81 M allocations: 123.843 MiB, 27.35% gc time)
  2.309175 seconds (5.96 M allocations: 264.596 MiB, 7.86% gc time)
```

This PR:
```
Benchmarking bab2.mps.gz
  9.079040 seconds (62.68 M allocations: 3.439 GiB, 8.40% gc time)
  3.562574 seconds (12.48 M allocations: 719.199 MiB, 10.11% gc time)
  7.591304 seconds (22.68 M allocations: 899.688 MiB, 18.04% gc time)
Benchmarking irish-electricity.mps.gz
  1.903456 seconds (17.94 M allocations: 1016.366 MiB, 10.26% gc time)
  1.697164 seconds (6.89 M allocations: 366.715 MiB, 6.05% gc time)
  1.947160 seconds (6.44 M allocations: 275.097 MiB, 7.50% gc time)
Benchmarking n2seq36q.mps.gz
  1.237586 seconds (5.54 M allocations: 365.401 MiB, 7.37% gc time)
  0.025388 seconds (101.24 k allocations: 17.046 MiB)
  0.405412 seconds (1.26 M allocations: 58.136 MiB)
Benchmarking rocI-4-11.mps.gz
  0.217302 seconds (1.32 M allocations: 77.872 MiB, 11.74% gc time)
  0.960789 seconds (3.81 M allocations: 192.257 MiB, 3.16% gc time)
  0.153749 seconds (616.85 k allocations: 26.944 MiB)
Benchmarking rocII-5-11.mps.gz
  1.421391 seconds (9.38 M allocations: 566.467 MiB, 6.63% gc time)
  0.133117 seconds (390.93 k allocations: 30.659 MiB, 52.01% gc time)
  0.586999 seconds (1.67 M allocations: 72.313 MiB)
Benchmarking roi5alpha10n8.mps.gz
  8.862825 seconds (64.08 M allocations: 3.681 GiB, 11.03% gc time)
  0.246268 seconds (463.14 k allocations: 119.874 MiB, 13.71% gc time)
  5.649430 seconds (10.38 M allocations: 479.186 MiB, 5.07% gc time)
Benchmarking supportcase42.mps.gz
  1.672369 seconds (12.48 M allocations: 748.771 MiB, 16.50% gc time)
  0.065130 seconds (306.58 k allocations: 33.658 MiB, 17.26% gc time)
  0.949660 seconds (2.75 M allocations: 110.861 MiB, 6.94% gc time)
Benchmarking timtab1.mps.gz
  0.004307 seconds (31.24 k allocations: 1.829 MiB)
  0.000803 seconds (4.40 k allocations: 457.641 KiB)
  0.004102 seconds (28.23 k allocations: 1.382 MiB)
Benchmarking triptim1.mps.gz
  2.076799 seconds (14.61 M allocations: 857.471 MiB, 9.49% gc time)
  0.084924 seconds (333.39 k allocations: 45.152 MiB, 16.72% gc time)
  1.130452 seconds (2.53 M allocations: 111.423 MiB, 5.42% gc time)
Benchmarking uccase12.mps.gz
  2.975312 seconds (16.54 M allocations: 986.554 MiB, 16.03% gc time)
  0.425015 seconds (1.79 M allocations: 123.502 MiB, 22.87% gc time)
  1.952220 seconds (5.96 M allocations: 264.596 MiB, 7.96% gc time)
```

**Comment by odow** [comment]
Now we're down to:
```
Benchmarking bab2.mps.gz
  6.268524 seconds (33.60 M allocations: 1.857 GiB, 9.00% gc time)
  0.524663 seconds (1.41 M allocations: 167.320 MiB, 19.75% gc time)
  5.658225 seconds (23.42 M allocations: 886.163 MiB, 16.26% gc time)
Benchmarking irish-electricity.mps.gz
  1.351090 seconds (10.74 M allocations: 608.127 MiB, 21.39% gc time)
  0.311046 seconds (1.56 M allocations: 101.770 MiB, 13.84% gc time)
  1.743961 seconds (6.52 M allocations: 281.954 MiB, 9.96% gc time)
Benchmarking n2seq36q.mps.gz
  0.921493 seconds (3.23 M allocations: 220.911 MiB, 3.40% gc time)
  0.053078 seconds (213.72 k allocations: 29.793 MiB)
  0.434221 seconds (2.04 M allocations: 91.360 MiB, 5.85% gc time)
Benchmarking rocI-4-11.mps.gz
  0.135539 seconds (875.09 k allocations: 50.561 MiB, 4.76% gc time)
  0.024456 seconds (193.60 k allocations: 12.111 MiB)
  0.124394 seconds (774.23 k allocations: 32.935 MiB, 7.57% gc time)
Benchmarking rocII-5-11.mps.gz
  0.963992 seconds (5.53 M allocations: 333.847 MiB, 4.73% gc time)
  0.085438 seconds (441.66 k allocations: 40.952 MiB, 13.83% gc time)
  0.617808 seconds (2.07 M allocations: 92.228 MiB, 8.54% gc time)
Benchmarking roi5alpha10n8.mps.gz
  6.581008 seconds (35.60 M allocations: 2.070 GiB, 7.09% gc time)
  0.336161 seconds (910.19 k allocations: 151.429 MiB)
  6.459736 seconds (14.09 M allocations: 629.152 MiB, 13.67% gc time)
Benchmarking supportcase42.mps.gz
  1.086017 seconds (6.93 M allocations: 432.418 MiB, 5.15% gc time)
  0.060734 seconds (298.55 k allocations: 33.200 MiB, 14.42% gc time)
  0.792056 seconds (2.75 M allocations: 112.805 MiB, 5.48% gc time)
Benchmarking timtab1.mps.gz
  0.004178 seconds (24.93 k allocations: 1.433 MiB)
  0.000935 seconds (5.01 k allocations: 473.047 KiB)
  0.005042 seconds (30.23 k allocations: 1.477 MiB)
Benchmarking triptim1.mps.gz
  1.449498 seconds (8.32 M allocations: 492.882 MiB, 5.11% gc time)
  0.104882 seconds (445.18 k allocations: 48.042 MiB, 17.28% gc time)
  1.039484 seconds (3.24 M allocations: 137.639 MiB, 5.73% gc time)
Benchmarking uccase12.mps.gz
  2.206237 seconds (10.31 M allocations: 617.703 MiB, 9.71% gc time)
  0.548663 seconds (1.85 M allocations: 125.201 MiB, 34.39% gc time)
  1.856745 seconds (6.29 M allocations: 281.404 MiB, 11.72% gc time)
```

**Comment by odow** [comment]
This is in a reasonable state now, and much better than before. I'm not planning on future changes. I'll leave open for a few days and then, barring objections, merge.
